src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -657,6 +697,72 @@ int getBufferedLength() {
             return (ibuffer == null ? 0 : ibuffer.size());
         }
 
+        /**
+         *  ByteBuffer wrapper for intrinsic implGCMCrypt
+         */
+        int implGCMCrypt(GCMOperation op, ByteBuffer src, ByteBuffer dst) {
+            int srcLen = src.remaining() - (src.remaining() % blockSize);
+            int rlen = srcLen;
+
+
+            if (srcLen < blockSize) {
+                return 0;
+            }
+
+            // 'in' and 'out' are always set together, just need to check 'in'
+            if (in == null || (in.length != PARALLEL_LEN
+                && in.length < srcLen)) {
+                in = new byte[Math.min(PARALLEL_LEN, srcLen)];
+                out = new byte[Math.min(PARALLEL_LEN, srcLen)];
+            }
+
+            byte[] ct;
+            if (encryption) {
+                ct = out;
+            } else {
+                ct = in;
+            }
+
+            if (srcLen > PARALLEL_LEN) {
+
+                if (src.hasArray() && dst.hasArray()) {
+                    int l = rlen;
+
+                    if (encryption) {
+                        rlen -= GaloisCounterMode.implGCMCrypt(src.array(),
+                            src.arrayOffset() + src.position(), src.remaining(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            gctr, ghash);
+                    } else {
+                        rlen -= GaloisCounterMode.implGCMCrypt(src.array(),
+                            src.arrayOffset() + src.position(), src.remaining(),
+                            src.array(), src.arrayOffset() + src.position(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            gctr, ghash);
+                    }
+                    src.position(src.position() + srcLen);
+                    dst.position(dst.position() + srcLen);
+                } else {
+
+                    do {
+                        src.get(in, 0, PARALLEL_LEN);
+                        rlen -= GaloisCounterMode.implGCMCrypt(in, 0,
+                            PARALLEL_LEN, ct, 0, out, 0, gctr, ghash);
+                        dst.put(out, 0, PARALLEL_LEN);
+                    } while (rlen > PARALLEL_LEN);
+                }
+            }
+            if (rlen >= blockSize) {
+                src.get(in, 0, rlen);
+                rlen = op.update(in, 0, rlen, out, 0);
+            }
+
+            dst.put(out, 0, rlen);
```

Comment:
This looks belong to the above if-block? I wonder how this have not affected the operation to fail. Perhaps the existing regression tests did not cover the 'rlen < blockSize' case. If the code in the above if-block is not run, this outsize dst.put(...) call would put extra output bytes into the output buffer.

Reply:
Yes... this one and the ct offset problem earlier I would have expected the regression test it pick the mistake.  There should be tests that catch this.. I'm not sure what's up.

Reply:
This shall be addressed in next update I assume?